In [1]:
import numpy as np

1. Создайте массив из 12 чисел, распределённых равномерно в диапазоне [-5, 5].

In [4]:
a1 = np.linspace(-5, 5, 12)
print(a1)

[-5.         -4.09090909 -3.18181818 -2.27272727 -1.36363636 -0.45454545
  0.45454545  1.36363636  2.27272727  3.18181818  4.09090909  5.        ]


2. Создайте массив 3×3 со значениями от 0 до 8.

In [5]:
a2 = np.arange(9).reshape(3, 3)
print(a2)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


3. Создайте массив из 10 чисел, начиная с 5 с шагом 3.

In [6]:
a3 = np.arange(5, 5 + 10 * 3, 3)
print(a3)

[ 5  8 11 14 17 20 23 26 29 32]


4. Найдите дисперсию и стандартное отклонение массива [2, 4, 6, 8, 10].

In [9]:
arr4 = np.array([2, 4, 6, 8, 10])
print("дисперсия =", arr4.var(), ", стандартное отклонение =", arr4.std())

дисперсия = 8.0 , стандартное отклонение = 2.8284271247461903


5. Создайте матрицу 3×3, найдите сумму её элементов по диагонали.

In [11]:
m5 = np.arange(1, 10).reshape(3, 3)
print("матрица\n", m5)
print("сумма по главной диагонали =", np.trace(m5))

матрица
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
сумма по главной диагонали = 15


6. Объедините три массива в один одномерный.

In [12]:
b1 = np.array([1, 2, 3])
b2 = np.array([4, 5, 6])
b3 = np.array([7, 8, 9])
merged = np.concatenate([b1, b2, b3])
print(merged)

[1 2 3 4 5 6 7 8 9]


7. Создайте массив 5×5 и вычислите кумулятивную сумму по строкам.

In [15]:
m7 = np.arange(1, 26).reshape(5, 5)
sum_rows = np.cumsum(m7, axis=1)
print("исходная матрица\n", m7)
print("   кумулятивная сумма по строкам\n", sum_rows)

исходная матрица
 [[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]
 [21 22 23 24 25]]
   кумулятивная сумма по строкам
 [[  1   3   6  10  15]
 [  6  13  21  30  40]
 [ 11  23  36  50  65]
 [ 16  33  51  70  90]
 [ 21  43  66  90 115]]


8. Из массива [1, 2, 3, 4, 5, 6] получите массив [1, 4, 9, 16, 25, 36] без циклов.

In [16]:
arr8 = np.arange(1, 7)
squares = arr8 ** 2
print(squares)

[ 1  4  9 16 25 36]


9. Замените все NaN в массиве [1, np.nan, 3, np.nan, 5] на 0.

In [18]:
arr9 = np.array([1, np.nan, 3, np.nan, 5])
cleaned = np.nan_to_num(arr9, nan=0)
print(cleaned)

[1. 0. 3. 0. 5.]


10. Создайте матрицу 4×4 и поверните её на 90 градусов.

In [19]:
m10 = np.arange(1, 17).reshape(4, 4)
rotated = np.rot90(m10)
print("исходная\n", m10)
print("повёрнутая на 90°\n", rotated)

исходная
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]
повёрнутая на 90°
 [[ 4  8 12 16]
 [ 3  7 11 15]
 [ 2  6 10 14]
 [ 1  5  9 13]]


2.1. вычислить значения на интервале [0, 1] с шагом 0.01 с точностью 0.01 вывести только положительные результаты F(x) = - cos x + 1

In [19]:
ar21 = np.arange(0.0, 1.0, 0.01)
ar21 = np.cos(ar21)
ar21 = np.multiply(ar21, -1)
ar21 = np.add(ar21, 1)
ar21 = np.round(ar21, 2)
ar21

array([0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ,
       0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.02, 0.02, 0.02, 0.02,
       0.02, 0.03, 0.03, 0.03, 0.03, 0.04, 0.04, 0.04, 0.04, 0.05, 0.05,
       0.05, 0.06, 0.06, 0.06, 0.07, 0.07, 0.08, 0.08, 0.08, 0.09, 0.09,
       0.1 , 0.1 , 0.1 , 0.11, 0.11, 0.12, 0.12, 0.13, 0.13, 0.14, 0.14,
       0.15, 0.15, 0.16, 0.16, 0.17, 0.17, 0.18, 0.19, 0.19, 0.2 , 0.2 ,
       0.21, 0.22, 0.22, 0.23, 0.24, 0.24, 0.25, 0.25, 0.26, 0.27, 0.28,
       0.28, 0.29, 0.3 , 0.3 , 0.31, 0.32, 0.33, 0.33, 0.34, 0.35, 0.36,
       0.36, 0.37, 0.38, 0.39, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.43, 0.44,
       0.45])

3.1. создать массив на основе ar21

In [10]:
ar22 = np.copy(ar21)
ar22 = np.round(ar22)
ar22

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

создать массив по размеру ar22, заполнить его интервальными значениями

In [16]:
ar23 = np.linspace(1, 200, len(ar21))
ar23 = np.round(ar23, 2)
ar23

array([  1.  ,   3.01,   5.02,   7.03,   9.04,  11.05,  13.06,  15.07,
        17.08,  19.09,  21.1 ,  23.11,  25.12,  27.13,  29.14,  31.15,
        33.16,  35.17,  37.18,  39.19,  41.2 ,  43.21,  45.22,  47.23,
        49.24,  51.25,  53.26,  55.27,  57.28,  59.29,  61.3 ,  63.31,
        65.32,  67.33,  69.34,  71.35,  73.36,  75.37,  77.38,  79.39,
        81.4 ,  83.41,  85.42,  87.43,  89.44,  91.45,  93.46,  95.47,
        97.48,  99.49, 101.51, 103.52, 105.53, 107.54, 109.55, 111.56,
       113.57, 115.58, 117.59, 119.6 , 121.61, 123.62, 125.63, 127.64,
       129.65, 131.66, 133.67, 135.68, 137.69, 139.7 , 141.71, 143.72,
       145.73, 147.74, 149.75, 151.76, 153.77, 155.78, 157.79, 159.8 ,
       161.81, 163.82, 165.83, 167.84, 169.85, 171.86, 173.87, 175.88,
       177.89, 179.9 , 181.91, 183.92, 185.93, 187.94, 189.95, 191.96,
       193.97, 195.98, 197.99, 200.  ])

создать массив из случайных элементов

In [23]:
ar24 = np.random.rand(len(ar21))*300
ar24 = np.round(ar24, 2)
ar24

array([168.39, 180.93, 204.8 , 212.5 , 240.5 , 107.45, 166.69,  14.63,
       194.92, 267.93, 289.69,   6.48, 200.37, 122.61, 224.55,   2.43,
        13.22, 181.56, 106.63,  75.01, 156.71, 124.78, 265.41,  20.13,
        49.17, 217.73, 244.56,  72.12, 121.98, 154.34,   5.23,   7.8 ,
       203.82, 239.96, 260.91, 166.32,  79.71, 242.66,   8.38, 264.3 ,
       202.79, 115.98, 264.57,  19.56,  94.24, 292.93,   7.55,  26.3 ,
        48.81, 256.21, 100.11,  96.82, 164.31,  17.24, 145.89, 103.84,
        25.9 , 158.89,  12.45,  57.25,  50.98, 156.5 , 260.56, 144.56,
       195.43, 289.46,  99.02,  80.74, 257.29,  34.34, 146.84,  68.16,
        44.91,  80.99, 153.78, 194.42, 289.22, 198.4 , 104.71, 131.57,
        57.42, 170.03,  47.35, 194.  , 241.72,  41.67, 161.7 , 267.66,
       243.02, 104.35, 150.82, 192.61,  40.31, 236.33, 128.31, 232.86,
        54.21,  62.01, 292.6 , 242.38])

детальное описание функции

In [25]:
np.info(np.random.rand)

 rand(*args)

rand(d0, d1, ..., dn)

Random values in a given shape.

.. note::
    This is a convenience function for users porting code from Matlab,
    and wraps `random_sample`. That function takes a
    tuple to specify the size of the output, which is consistent with
    other NumPy functions like `numpy.zeros` and `numpy.ones`.

Create an array of the given shape and populate it with
random samples from a uniform distribution
over ``[0, 1)``.

Parameters
----------
d0, d1, ..., dn : int, optional
    The dimensions of the returned array, must be non-negative.
    If no argument is given a single Python float is returned.

Returns
-------
out : ndarray, shape ``(d0, d1, ..., dn)``
    Random values.

See Also
--------
random

Examples
--------
>>> np.random.rand(3,2)
array([[ 0.14022471,  0.96360618],  #random
       [ 0.37601032,  0.25528411],  #random
       [ 0.49313049,  0.94909878]]) #random


векторные операции

In [30]:
ar25 = np.add(ar21, ar22)
ar25 = np.add(ar25, ar23)
ar25 = np.add(ar25, ar24)
ar25 = np.multiply(ar25, 5)
ar25

array([ 841.95,  904.65, 1024.  , 1062.5 , 1202.5 ,  537.25,  833.45,
         73.15,  974.6 , 1339.65, 1448.45,   32.5 , 1001.95,  613.15,
       1122.85,   12.25,   66.2 ,  907.9 ,  533.35,  375.25,  783.75,
        624.1 , 1327.25,  100.95,  246.15, 1088.95, 1223.1 ,  361.  ,
        610.3 ,  772.1 ,   26.55,   39.5 , 1019.6 , 1200.3 , 1305.15,
        832.2 ,  399.15, 1214.  ,   42.6 , 1322.3 , 1014.75,  580.7 ,
       1323.75,   98.7 ,  472.2 , 1465.65,   38.75,  132.6 ,  245.15,
       1282.25,  501.75,  485.4 ,  822.85,   87.6 ,  730.85,  520.7 ,
        131.  ,  796.05,   63.85,  287.95,  256.6 ,  784.3 , 1304.7 ,
        724.7 ,  979.15, 1449.3 ,  497.2 ,  405.9 , 1288.65,  174.  ,
        736.6 ,  343.2 ,  227.05,  407.45,  771.5 ,  974.8 , 1448.9 ,
        994.8 ,  526.45,  660.85,  290.1 ,  853.25,  239.95,  973.3 ,
       1211.9 ,  211.75,  812.  , 1341.9 , 1218.7 ,  525.45,  757.9 ,
        966.95,  205.45, 1185.65,  645.65, 1168.5 ,  275.35,  314.35,
       1467.4 , 1216

In [31]:
np.subtract(ar21, ar22)

array([0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ,
       0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.02, 0.02, 0.02, 0.02,
       0.02, 0.03, 0.03, 0.03, 0.03, 0.04, 0.04, 0.04, 0.04, 0.05, 0.05,
       0.05, 0.06, 0.06, 0.06, 0.07, 0.07, 0.08, 0.08, 0.08, 0.09, 0.09,
       0.1 , 0.1 , 0.1 , 0.11, 0.11, 0.12, 0.12, 0.13, 0.13, 0.14, 0.14,
       0.15, 0.15, 0.16, 0.16, 0.17, 0.17, 0.18, 0.19, 0.19, 0.2 , 0.2 ,
       0.21, 0.22, 0.22, 0.23, 0.24, 0.24, 0.25, 0.25, 0.26, 0.27, 0.28,
       0.28, 0.29, 0.3 , 0.3 , 0.31, 0.32, 0.33, 0.33, 0.34, 0.35, 0.36,
       0.36, 0.37, 0.38, 0.39, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.43, 0.44,
       0.45])

In [34]:
np.round(np.multiply(ar21, ar23), 2)

array([0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ,
       0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ,
       0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ,
       0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.01, 0.01, 0.01, 0.01, 0.01,
       0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.02, 0.02, 0.02, 0.02,
       0.02, 0.02, 0.03, 0.03, 0.03, 0.03, 0.03, 0.04, 0.04, 0.04, 0.04,
       0.04, 0.05, 0.05, 0.05, 0.06, 0.06, 0.06, 0.06, 0.07, 0.07, 0.08,
       0.08, 0.08, 0.09, 0.09, 0.1 , 0.1 , 0.11, 0.11, 0.12, 0.12, 0.13,
       0.13, 0.14, 0.14, 0.15, 0.15, 0.16, 0.17, 0.18, 0.18, 0.18, 0.19,
       0.2 ])

In [38]:
np.round(np.divide(ar25, ar21), 2)

C:\Users\goblin\AppData\Local\Temp\ipykernel_14996\1862689696.py:1: RuntimeWarning: divide by zero encountered in divide
  np.round(np.divide(ar25, ar21), 2)


array([      inf,       inf,       inf,       inf,       inf,       inf,
             inf,       inf,       inf,       inf,       inf,   3250.  ,
       100195.  ,  61315.  , 112285.  ,   1225.  ,   6620.  ,  90790.  ,
        26667.5 ,  18762.5 ,  39187.5 ,  31205.  ,  66362.5 ,   3365.  ,
         8205.  ,  36298.33,  40770.  ,   9025.  ,  15257.5 ,  19302.5 ,
          663.75,    790.  ,  20392.  ,  24006.  ,  21752.5 ,  13870.  ,
         6652.5 ,  17342.86,    608.57,  16528.75,  12684.38,   7258.75,
        14708.33,   1096.67,   4722.  ,  14656.5 ,    387.5 ,   1205.45,
         2228.64,  10685.42,   4181.25,   3733.85,   6329.62,    625.71,
         5220.36,   3471.33,    873.33,   4975.31,    399.06,   1693.82,
         1509.41,   4357.22,   6866.84,   3814.21,   4895.75,   7246.5 ,
         2367.62,   1845.  ,   5857.5 ,    756.52,   3069.17,   1430.  ,
          908.2 ,   1629.8 ,   2967.31,   3610.37,   5174.64,   3552.86,
         1815.34,   2202.83,    967.  ,   2752.42, 

In [39]:
np.round(np.power(ar21, ar23), 2)

array([1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  ,
       0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.92, 0.92, 0.92, 0.92,
       0.92, 0.9 , 0.9 , 0.9 , 0.9 , 0.88, 0.88, 0.88, 0.88, 0.86, 0.86,
       0.86, 0.84, 0.84, 0.84, 0.83, 0.83, 0.82, 0.82, 0.82, 0.81, 0.81,
       0.79, 0.79, 0.79, 0.78, 0.78, 0.78, 0.78, 0.77, 0.77, 0.76, 0.76,
       0.75, 0.75, 0.75, 0.75, 0.74, 0.74, 0.73, 0.73, 0.73, 0.72, 0.72,
       0.72, 0.72, 0.72, 0.71, 0.71, 0.71, 0.71, 0.71, 0.7 , 0.7 , 0.7 ,
       0.7 , 0.7 , 0.7 , 0.7 , 0.7 , 0.69, 0.69, 0.69, 0.69, 0.69, 0.69,
       0.69, 0.69, 0.69, 0.69, 0.69, 0.69, 0.69, 0.69, 0.7 , 0.7 , 0.7 ,
       0.7 ])

3.2 МКБ

На основе данных об изменении стоимости активов рассчитайте
экономико-статистические показатели c целью дальнейшей оценки
57
рисков инвестирования: минимальное значение; максимальное значение; математическое ожидание; дисперсию; стандартное отклонение;
среднеквадратическое отклонение и коэффициент вариации.
Для выполнения данного задания необходимо выполнить ниже
перечисленные действия:
1. создать массив на основании официальных статистических
данных о курсе акций за предыдущий месяц (тип данных массива –
str, источник данных – https://ru.investing.com/equities/, эмитент акций – по вариантам, применить функции (на выбор): «numpy.loadtxt»,
«numpy.fromfile»);
2. создать массив, который должен содержать только информацию о ценах за каждый торговый день: ценах открытия, минимальных
ценах, максимальных ценах и ценах закрытия (тип данных массива –
float, источник данных – созданный в п.1 задания массив, эмитент акций – по вариантам, применить функции (на выбор): «numpy.split»,
«numpy.split», «numpy.hsplit», использование среза совместно с
«numpy.array»);
3. создать массив (вектор-столбец), который должен содержать
информацию о средней цене актива за каждый торговый день (применить функции (совместно): «numpy.mean», «numpy.Reshape»);
4. осуществить слияние массивов, созданных на шаге 2 (слева) и
шаге 3 (справа) (применить функции (на выбор): «numpy.concatenate»,
«numpy.hstack»);
5. рассчитать минимальное значение, максимальное значение,
математическое ожидание, дисперсию, стандартное отклонение,
среднеквадратическое отклонение и коэффициент вариации по всем
видам цен активов. Для расчетов показателей использовать следующие функции и способы расчета:
- математическое ожидание («numpy.mean»);
- дисперсия («numpy.mean»);
- стандартное отклонение («numpy.std»);
- среднеквадратическое отклонение (корень квадратный из дисперсии);
- коэффициент вариации (среднеквадратическое отклонение/математическое ожидание));
6. запишите полученный массив на диск (применить функцию:
«numpy.save»)